crates/bitwarden-crypto/src/xchacha20.rs

```diff
@@ -0,0 +1,137 @@
+//! # XChaCha20Poly1305 operations
+//!
+//! Contains low level XChaCha20Poly1305 operations used by the rest of the crate.
+//!
+//! In most cases you should use the [EncString][crate::EncString] with
+//! [KeyEncryptable][crate::KeyEncryptable] & [KeyDecryptable][crate::KeyDecryptable] instead.
+
+use chacha20poly1305::{AeadCore, AeadInPlace, KeyInit, XChaCha20Poly1305};
+use generic_array::{typenum::U24, GenericArray};
+
+/**
+ * Note:
+ * XChaCha20Poly1305 encrypts data, and authenticates both the cipher text and associated
+ * data. This does not provide key-commitment, and assumes there can only be one key.
+ *
+ * If multiple keys are possible, a key-committing cipher such as
+ * XChaCha20Poly1305Blake3CTX should be used: `https://github.com/bitwarden/sdk-internal/pull/41` to prevent invisible-salamander style attacks.
+ * `https://eprint.iacr.org/2019/016.pdf`
+ * `https://soatok.blog/2024/09/10/invisible-salamanders-are-not-what-you-think/`
+ */
+use crate::CryptoError;
+
+#[allow(unused)]
+pub(crate) fn generate_nonce() -> GenericArray<u8, U24> {
+    XChaCha20Poly1305::generate_nonce(rand::thread_rng())
+}
+
+#[allow(unused)]
+pub(crate) fn encrypt_xchacha20_poly1305(
+    nonce: &[u8; 24],
```

Comment:
On the `aes` module we generate the `IV` internally to ensure we always generate a random one each time for encryption, assuming that the `nonce` works the same in `chacha20`, have you considered not exposing this as a parameter for encryption and just generating it internally?

Reply:
As mentioned for the nonce generating function, initially I thought the IV needed to be authenticated, but this is not the case, so I'll move this to generate during encrypt.